transport/http/src/main/java/org/kaazing/gateway/transport/http/HttpAcceptor.java

```diff
@@ -593,6 +610,80 @@ public void removeBridgeFilters(IoFilterChain filterChain) {
         }
     }
 
+    // If HTTP's X-FORWARDED-FOR header is set, then use it's value to determine the remote
+    // IP address. If X-FORWARDED-FOR header is NOT set, then use IoSession to determine
+    // the remote IP address.
+    private String getRemoteIpAddress(IoSession session, HttpRequestMessage httpRequest) {
+        String remoteIpAddr = httpRequest.getHeader(HEADER_X_FORWARDED_FOR);
+
+        if (remoteIpAddr == null) {
+            SocketAddress socketAddr = session.getRemoteAddress();
+
+            if (session instanceof AbstractBridgeSession) {
+                IoSession tcpSession = getTcpSession(session);
+                if (tcpSession == null) {
+                    return null;
+                }
+                socketAddr = tcpSession.getRemoteAddress();
+            }
+
+            InetAddress inetAddr = ((InetSocketAddress)socketAddr).getAddress();
+            remoteIpAddr = inetAddr.getHostAddress();
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("Using IoSession to determine remote IP address: [%s].", remoteIpAddr);
+            }
+        }
+        else {
+            // X-FORWARDED-FOR HTTP header is inserted by proxies to identify the remote IP address.
+            // If the proxies are chained, then X-FORWARDED-FOR header can contain multiple IP
+            // addresses separated by commas. In such a scenario, the first IP address represents
+            // the remote IP address.
+            int index = remoteIpAddr.indexOf(',');
+            if (index > -1) {
+                remoteIpAddr = remoteIpAddr.substring(0, index).trim();
+            }
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("Using X-FORWARDED-FOR header value for remote IP address: [%s].", remoteIpAddr);
+            }
+        }
+
+        return remoteIpAddr;
+    }
+
+    private IoSession getTcpSession(IoSession session) {
+    IoSession tempSession = session;
+
+    while (tempSession != null) {
+            if (tempSession instanceof NioSocketChannelIoSession) {
+                return tempSession;
+            }
+
+            if (!(tempSession instanceof AbstractBridgeSession<?, ?>)) {
+                return null;
+            }
+
+            AbstractBridgeSession<?, ?> bridgeSession = (AbstractBridgeSession<?, ?>) tempSession;
+            tempSession = bridgeSession.getParent();
+        }
+
+        return null;
+    }
+
+    private static HttpSubjectSecurityFilter getSubjectSecurityFilter(Logger logger) {
+        Class<HttpSubjectSecurityFilter> clazz = HttpSubjectSecurityFilter.class;
+        ServiceLoader<HttpSubjectSecurityFilter> loader = ServiceLoader.load(clazz);
+
+        if (loader != null) {
+            if ((loader.iterator() != null) && loader.iterator().hasNext()) {
+                return loader.iterator().next();
```

Comment:
If the filter is found by service loader, how is the filter get hold of given logger ?